core/src/Kokkos_HPX.hpp

```diff
@@ -49,124 +49,76 @@ static_assert(false,
-// shared data early, as (the pointer to) the shared data will not be used
-// anymore by the closure at the end of execute_task.
-//
-// We also mark the shared instance data as mutable so that we can reset it
-// from the const execute_task member function.
-
 namespace Kokkos {
 namespace Impl {
-class thread_buffer {
+class hpx_thread_buffer {
   static constexpr std::size_t m_cache_line_size = 64;
 
-  std::size_t m_num_threads;
-  std::size_t m_size_per_thread;
-  std::size_t m_size_total;
-  char *m_data;
+  std::size_t m_num_threads     = 0;
+  std::size_t m_size_per_thread = 0;
+  std::size_t m_extra_space     = 0;
+  std::size_t m_size_total      = 0;
+  char *m_data                  = nullptr;
 
-  void pad_to_cache_line(std::size_t &size) {
+  static constexpr void pad_to_cache_line(std::size_t &size) {
     size = ((size + m_cache_line_size - 1) / m_cache_line_size) *
            m_cache_line_size;
   }
 
  public:
-  thread_buffer()
-      : m_num_threads(0),
-        m_size_per_thread(0),
-        m_size_total(0),
-        m_data(nullptr) {}
-  thread_buffer(const std::size_t num_threads,
-                const std::size_t size_per_thread) {
-    resize(num_threads, size_per_thread);
-  }
-  ~thread_buffer() { delete[] m_data; }
-
-  thread_buffer(const thread_buffer &) = delete;
-  thread_buffer(thread_buffer &&)      = delete;
-  thread_buffer &operator=(const thread_buffer &) = delete;
-  thread_buffer &operator=(thread_buffer) = delete;
-
-  void resize(const std::size_t num_threads,
-              const std::size_t size_per_thread) {
-    m_num_threads     = num_threads;
-    m_size_per_thread = size_per_thread;
-
-    pad_to_cache_line(m_size_per_thread);
-
-    std::size_t size_total_new = m_num_threads * m_size_per_thread;
-
-    if (m_size_total < size_total_new) {
-      delete[] m_data;
-      m_data       = new char[size_total_new];
-      m_size_total = size_total_new;
-    }
-  }
-
-  char *get(std::size_t thread_num) {
-    assert(thread_num < m_num_threads);
-    if (m_data == nullptr) {
-      return nullptr;
-    }
-    return &m_data[thread_num * m_size_per_thread];
-  }
+  hpx_thread_buffer() = default;
+  ~hpx_thread_buffer() { delete[] m_data; }
+  hpx_thread_buffer(const hpx_thread_buffer &) = delete;
+  hpx_thread_buffer(hpx_thread_buffer &&)      = delete;
+  hpx_thread_buffer &operator=(const hpx_thread_buffer &) = delete;
+  hpx_thread_buffer &operator=(hpx_thread_buffer) = delete;
+
+  void resize(const std::size_t num_threads, const std::size_t size_per_thread,
+              const std::size_t extra_space = 0) noexcept;
+  void *get(std::size_t thread_num) const noexcept;
+  void *get_extra_space() const noexcept;
+};
 
-  std::size_t size_per_thread() const noexcept { return m_size_per_thread; }
-  std::size_t size_total() const noexcept { return m_size_total; }
+template <typename T>
+struct hpx_range {
+  const T begin;
+  const T end;
```

Comment:
Why `const`?  This implicitly deletes the assignment operators and makes it harder to reason about that type.

Reply:
No strong reason, other than that they don't need to be changed after construction. Happy to un-const them if you prefer that.

Reply:
Changed in https://github.com/kokkos/kokkos/pull/5628/commits/2ecb963428b7f814d43cd413d7fcd42ec43c530e.